### PR TITLE
💡 Lab Project: Monte Carlo 🏁

### PR DESCRIPTION
**Pull request by warrenbuckley**
# 🏁 Project Monte Carlo 🏁 
This PR implements the code editor engine behind **VS Code** that is open sourced and called **Monaco**
https://microsoft.github.io/monaco-editor

_This is a working proof of concept PR to experiment in and see if we can get a better experience than the current code editor of ACE._

# NOTE: This is a lab project and is a draft PR, this still needs to be discussed and considered if it will ever make it back into the core CMS first

## JavaScript Editor
* Powerful intellisense
* Find unused code

monaco-scripts-editor
_Click GIF to enlarge_

## CSS Editor
* Powerful intellisense
* More upto date CSS properties for intellisense than ACE
* Nice color picker
* Custom intellisense with a snippet insert with TAB stops to help fill out RTE styles in the code editor

monaco-css-editor
_Click GIF to enlarge_

## Grid Settings & Styles JSON
* No longer just a textarea
* Intellisense & validation by using a JSON schema to specify shape of JSON

monaco-grid-json-editor
_Click GIF to enlarge_

## TODOs
- [x] RTE HTML Editor
- [x] Template Editor
- [x] Partial View Editor
- [x] Partial Macro View Editor
- [x] Insert Value button - adds content to editor
- [x] Layout button - adds content to editor
- [x] Sections button - adds content to editor
- [x] Insert Query - adds content to editor
- [x] Keyboard shortcuts to open dialog for insert macro etc
- [x] Completion of `model.` - use WebAPI to determine which ModelsBuilder properties to return
- [ ] ~Try to change color of Razor tag items from red to yellow background similar to ACE & VS~
- [ ] Auto indent / code format 
- [ ] Completion of `layout = ` - use WebAPI to fetch a list of views/templates to choose from
- [ ] Completion of `Umbraco.` - use WebAPI to fetch methods & properties on UmbracoHelper

### Completion Challenges
* Ensuring we trigger our WebAPI calls when necessary on correctly matched words
* Use reflection for Umbraco Helper to get methods & any extension methods devs have added
* Not sure how we can validate/verify the correct params are passed to a method like full VS
* Unsure we can get completion for items inside a for loop and var named item for example

## Resources
Blog about custom autocomplete for XML using an XSD Schema & how to customise Monaco
https://mono.software/2017/04/11/custom-intellisense-with-monaco-editor/

Using ESLint & Prettier in Monaco editor for JS files
https://blog.expo.io/building-a-code-editor-with-monaco-f84b3a06deaf

Tokenize & style custom language elements
https://medium.com/galvanize/building-a-low-effort-high-impact-prototype-for-hackfest-online-code-editor-for-aclscript-a2005bb9f83b

## Repos of interest
https://github.com/Microsoft/vscode
https://github.com/Microsoft/monaco-languages

https://github.com/Microsoft/monaco-css
https://github.com/microsoft/vscode-css-languageservice

https://github.com/Microsoft/monaco-html
https://github.com/Microsoft/vscode-html-languageservice

https://github.com/Microsoft/monaco-typescript

https://github.com/Microsoft/monaco-json
https://github.com/Microsoft/vscode-json-languageservice

## Idea: Sass Editor
* Monaco supports Sass as a language
* ClientDependancy from @Shazwazza has a Sass compiler extension

## Idea: Use Omnisharp to do completions, formatting etc
Use OmniSharp project as the LSP (Language Service Protocol)

**JS client to add LSP to Monaco**
https://www.typefox.io/blog/teaching-the-language-server-protocol-to-microsofts-monaco-editor
https://github.com/TypeFox/monaco-languageclient

**HTTP Server - no docs of usage though**
https://github.com/OmniSharp/omnisharp-roslyn#using-omnisharp

**OmniSharp NuGet packages to try & consume/experiment with**
https://www.nuget.org/packages?q=OmniSharp

Prototype of this working using TypeFox LSP client lib & Omnisharp but its using the executable as opposed to the HTTP server
https://github.com/patilarpith/monaco-languageclient-omnisharp-lsp

### CODE REVIEW

**Comment by warrenbuckley** [comment]
Yes originally in earlier versions of the hack this was an attribute and it was hard to tell what things was related & specific to the code editor when there are other unrelated things on the div at the time.

Good idea @bjarnef will fix that up :)

File: src/Umbraco.Web.UI.Client/package.json [hunk 1]
```diff
@@ -37,6 +37,7 @@
     "jquery-ui-touch-punch": "0.2.3",
     "lazyload-js": "1.0.0",
     "moment": "2.22.2",
+    "monaco-editor": "^0.19.3",
```
Comment (rasmuseeg) [inline src/Umbraco.Web.UI.Client/package.json]: Perhaps it should just lock to this version? :-)

File: src/Umbraco.Web.UI.Client/src/views/stylesheets/edit.controller.js [hunk 2]
```diff
@@ -171,78 +164,99 @@
...
-            vm.setDirty = function () {
-                setFormState("dirty");
+            // Options to pass to code editor (VS-Code)
+            vm.codeEditorOptions = {
+                language: "css"
             }
 
-            //sync state
-            editorState.set(vm.stylesheet);
+            //monaco.languages.registerCompletionItemProvider().dispose();
```
Comment (rasmuseeg) [inline src/Umbraco.Web.UI.Client/src/views/stylesheets/edit.controller.js]: Do we still need this? :-)

**Comment by warrenbuckley** [comment]
Thanks @rasmuseeg for the feedback 😍 
Keep your eyes peeled on this PR/branch as I progress.

File: src/Umbraco.Web.UI.Client/src/views/stylesheets/edit.controller.js (hunk 2 above)
Comment (warrenbuckley) [reply src/Umbraco.Web.UI.Client/src/views/stylesheets/edit.controller.js]: Good spot - it was my scratchpad for seeing what APIs are available in Monaco with their TypeScript definition file

**Comment by bjarnef** [comment]
@warrenbuckley you might also want to try using Monaco Editor as readonly mode inside the `umb-code-snippet` component.
https://github.com/umbraco/Umbraco-CMS/blob/d51b5e78221ea0c9658d0f5c47d3eefb0ffdcd99/src/Umbraco.Web.UI.Client/src/views/components/umb-code-snippet.html

At the moment this doesn't have specific syntax highlighting, but via the language parameter it could be possible to pass in the language: HTML, CSS, JS, C# etc. 
https://github.com/umbraco/Umbraco-CMS/blob/d51b5e78221ea0c9658d0f5c47d3eefb0ffdcd99/src/Umbraco.Web.UI.Client/src/common/directives/components/umbcodesnippet.directive.js#L50

File: src/Umbraco.Web.UI.Client/package.json (hunk 1 above)
Comment (warrenbuckley) [reply src/Umbraco.Web.UI.Client/package.json]: If this POC ever gets considered/accepted to be put in then, yes that's a good idea to lock the version

**Comment by warrenbuckley** [comment]
## Intellisense in Templates
I have managed to get some basic autocompletion/intellisense to work by using reflection to find the model used in the view. 

**Unfortunately there is no way to know what to complete if you were in a foreach loop etc**

monte-carlo_template-models
_Click GIF to enlarge_

### What next?
Well I have investigated into looking Roslyn & or OmniSharp for this prototype but have not succeeded, due to unable to **convert from 'System.Web.Razor.Parser.SyntaxTree.Block' to 'Microsoft.CodeAnalysis.Document'** with the attempted example below.

```csharp

        [HttpPost]
        public string FetchCompletions(RazorCode model)
        {
            var host = new System.Web.Razor.RazorEngineHost(new CSharpRazorCodeLanguage());
            var templateEngine = new System.Web.Razor.RazorTemplateEngine(host);
            
            var compiledCode = templateEngine.GenerateCode(new StringReader(model.Code));
            var parsed = templateEngine.ParseTemplate(new StringReader(model.Code));

            // Cannot convert from 'System.Web.Razor.Parser.SyntaxTree.Block' to 'Microsoft.CodeAnalysis.Document'
            var completionService = Microsoft.CodeAnalysis.Completion.CompletionService.GetService(compiledCode.Document);
            var completions = completionService.GetCompletionsAsync(compiledCode.Document, 1);

            return "hello warren";
        }
```

## Potential to do this in project UniCore

With project UniCore aka conversion to .NET Core the underlying legacy Razor engine will be updated to the newer version that supports things like TagHelpers. With my research & findings it looks more likely to be possible to use Roslyn CompletionService as it has the following extension method that would get a CodeAnalysis.Document from the Razor

https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.razor.language.razorcodedocumentextensions.getcsharpdocument?view=aspnetcore-3.0

**Pull request closed without merging**